### PR TITLE
Threading exp

### PR DESCRIPTION
**Pull request by Rav263**

### CODE REVIEW

File: generate_turns.py [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+
+def generate_turns_pown(pos, plate, possible_turns):
+    if plate.get_color_map(pos) == plate.white:
+        if pos[0] == 6 and plate.get_type_map((pos[0] - 2, pos[1])) == plate.empty_map:
+            possible_turns[(pos[0] - 2, pos[1])].append(pos)
+
+        if plate.get_type_map((pos[0] - 1, pos[1])) == plate.empty_map:
+            possible_turns[(pos[0] - 1, pos[1])].append(pos)
+    else:
+        if pos[0] == 1 and plate.get_type_map((pos[0] + 2, pos[1])) == plate.empty_map:
```
Comment (p0l0satik) [inline generate_turns.py:17]: константа 2

File: generate_turns.py [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+                if plate.get_color_map(turn_end) != color:
+                    possible_turns[turn_end].append(pos)
+                break
+
+            possible_turns[turn_end].append(pos)
+
+
+def generate_turns_queen(pos, plate, possible_turns, color):
+    generate_turns_rook(pos, plate, possible_turns, color)
+    generate_turns_rook(pos, plate, possible_turns, color)
```
Comment (p0l0satik) [inline generate_turns.py:76]: замени на слона

File: check_turn.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+import math_functions as mf
+
+
+def check_pown(turn, plate):
```
[the remark this answers is not in the document]
Comment (Rav263) [reply check_turn.py:4]: Changet

File: field.py [hunk 4]
```diff
@@ -1,20 +1,72 @@
+from itertools import product
+
+
 class Field:
-    def __init__(self, data):
-        print("Init field class")
-        self.load_field(data)
+    empty_map = 0
+    pown = 1
```
[the remark this answers is not in the document]
Comment (Rav263) [reply field.py:6]: This is stupid thing to do, but i did this

File: field.py [hunk 5]
```diff
@@ -1,20 +1,72 @@
...
 
-    def do_turn(self, turn):
-        print("Doing turn from {0} to {1}".format(turn.start_pos, turn.end_pos))
+    def __init__(self, data, copy=None):
+        if copy is None:
+            print("Init field class")
+            self.load_field(data)
+            self.field_size = data.field_size
+        else:
+            self.field = [x.copy() for x in copy.field]
```
[the remark this answers is not in the document]
Comment (Rav263) [reply field.py:21]: yes of course

[the remark this answers is not in the document]
Comment (Rav263) [reply field.py:21]: Done

File: field.py [hunk 6]
```diff
@@ -1,20 +1,72 @@
...
+        self.set_map(turn.start_pos, fig)
+        return tmp
+
+    def copy(self):
+        return Field(None, self)
 
     def load_field(self, data):
         self.field = data.data["FIELD"]
 
     def get_map(self, pos):
```
[the remark this answers is not in the document]
Comment (Rav263) [reply field.py:38]: Done and Done

File: gamelogic.py [hunk 7]
```diff
@@ -27,110 +69,177 @@ def start(self, plate, data):
...
+                    gt.generate_turns_king(pos, plate, possible_turns, color)
 
-            for i in range(sign_1, mod_1, sign_1):
-                if plate.get_map((turn.end_pos[0] + i, turn.end_pos[1] + i * sign_1)) != 0:
-                    return False
+        return possible_turns
 
-            return True
+    def thread_generate(self, plate, color, depth, turns, index, return_dict):
+        best_cost = -9999 if color == plate.black else 9999
```
[the remark this answers is not in the document]
Comment (Rav263) [reply gamelogic.py:158]: I hate constants

File: gamelogic.py [hunk 8]
```diff
@@ -12,13 +26,41 @@ def print(self):
...
         print("Init game logic class")
 
     def start(self, plate, data):
         color = 1
+        self.figures_cost = data.data["FIGURES_COST"]
 
+        global DATA
+        DATA = data
         while True:
             io_functions.print_field(plate.field, data)
```
[the remark this answers is not in the document]
Comment (Rav263) [reply gamelogic.py:65]: this is debug value i will remove it

File: gamelogic.py [hunk 9]
```diff
@@ -27,110 +69,177 @@ def start(self, plate, data):
...
+                                   args=(now_plate, color, depth, turns[start:end + 1], i, return_dict)))
+
+            threads[len(threads) - 1].start()
+
+        for thread in threads:
+            thread.join()
+
+        return max(return_dict.values(), key=lambda x: x[1])
+
+    def ai_turn(self, plate, color, depth, alpha=-10000, beta=10000):
```
[the remark this answers is not in the document]
Comment (Rav263) [reply gamelogic.py:213]: FIX

File: generate_turns.py (hunk 2 above)
Comment (Rav263) [reply generate_turns.py:76]: FIX

File: gamelogic.py [hunk 10]
```diff
@@ -12,13 +26,41 @@ def print(self):
...
+    return (int)(os.popen('grep -c cores /proc/cpuinfo').read())
+
+
 class Logic:
     def __init__(self):
         print("Init game logic class")
 
     def start(self, plate, data):
         color = 1
+        self.figures_cost = data.data["FIGURES_COST"]
```
[the remark this answers is not in the document]
Comment (Rav263) [reply gamelogic.py]: FIX

File: generate_turns.py [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+
+from itertools import permutations
+from itertools import islice
+from itertools import count
+from itertools import product
+
+
+def generate_turns_pown(pos, plate, possible_turns):
+    if plate.get_color_map(pos) == plate.white:
+        if pos[0] == 6 and plate.get_type_map((pos[0] - 2, pos[1])) == plate.empty_map:
```
[the remark this answers is not in the document]
Comment (Rav263) [reply generate_turns.py]: 2 is not constant

File: generate_turns.py (hunk 1 above)
Comment (Rav263) [reply generate_turns.py]: The same

**Pull request merged**